### PR TITLE
MODFISTO-63 implemented encumbrance restrictions logic

### PR DESCRIPTION
**Pull request by grabsefx**
## Purpose
https://issues.folio.org/browse/MODFISTO-63

## Approach
* implemented encumbrance restrictions logic

#### TODOS and Open Questions
* point acq-models to the latest master after merging https://github.com/folio-org/mod-finance-storage/pull/74

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 1]
```diff
@@ -67,4 +74,36 @@ private Transaction modifyTransaction(Transaction transaction, AwaitingPayment a
...
+    return checkEncumbranceRestrictions(encumbrance)
+      .thenCompose(v -> createTransaction(encumbrance));
+  }
+
+  public CompletableFuture<Void> checkEncumbranceRestrictions(Transaction encumbrance) {
+    String budgetQuery = "fundId==" + encumbrance.getFromFundId() + " AND fiscalYearId==" + encumbrance.getFiscalYearId();
+    return new BudgetsHelper(httpClient, okapiHeaders, ctx, lang).getSingleBudgetByQuery(budgetQuery)
+      .thenCompose(budget -> {
+        if (budget.getBudgetStatus() == Budget.BudgetStatus.ACTIVE) {
+          String ledgerQuery = "fund.id==" + encumbrance.getFromFundId() + " AND fiscalYear.id==" + encumbrance.getFiscalYearId();
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java]: Ledger not depends on fiscalYear, so you only need a fund.id to find the related ledger.

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 2]
```diff
@@ -67,4 +74,36 @@ private Transaction modifyTransaction(Transaction transaction, AwaitingPayment a
...
+        } else {
+          throw new HttpException(422, BUDGET_IS_INACTIVE);
+        }
+      });
+  }
+
+  private void checkLedgerEncumbranceRestrictions(Transaction encumbrance, Budget budget, Ledger ledger) {
+    if (ledger.getRestrictEncumbrance() && budget.getAllowableEncumbrance() != null) {
+      Double availableForEncumbrance = getBudgetAvailableForEncumbrance(budget.getAvailable(), budget.getAllowableEncumbrance(), encumbrance.getCurrency());
+      if (encumbrance.getAmount() > availableForEncumbrance) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java]: It is not clear from the WIKI how to calculate Remaining Allowable Encumbrance. @craigmcnally couldn't you shed some light on that?

Comment (craigmcnally) [reply src/main/java/org/folio/rest/helper/TransactionsHelper.java]: it's on the wiki...

> The "Remaining Allowable Enc." in the flowchar is calculated as follows:
>
>[remaining amount we can encumber] = (allocated * allowableEncumbered) - (allocated - (unavailable + available)) - (encumbered + awaitingPayment)
>
>The "restrictEncumbrances" flag actually lives in the ledger, so we'll need to find traverse:  budget → fund → ledger.restrictEncumbrances

**Comment by craigmcnally** [comment]
I see that MODFISTO-63 was implemented in mod-finance.  This is ok, but I think we need to check this in the storage layer.  Performing these checks in both places isn't a bad thing, so I'm definitely not suggesting we scrap this work. 

The reason it needs to happen in the storage layer is that the transactions pass through the business logic module and are persisted in a temporary table when they get to the DB.  They aren't applied until all encumbrances for that order are received.  There's no guarantee that something won't change between the time the transactions pass through the BL module and when all encumbrances are received and are then applied.

Side note:  I see in one of the commits above that MODFIN-63 is referenced.  I think this is probably a mistake.  If possible please ammend or squash your commits to prevent this commit message from confusing us in the future if we ever look back at this work

**Comment by grabsefx** [comment]
All checks moved to mod-finance-storage module https://github.com/folio-org/mod-finance-storage/pull/77
Another PR https://github.com/folio-org/mod-orders/pull/239 just for preparing appropriate error messages

Current PR is going to be closed.

**Pull request closed without merging**